### ISSUE TITLE
bash-exec reporting false failures

### ISSUE THREAD

**Issue by sabbene** [github/super-linter#2667]
## Describe the bug

bash-exec linter is reporting that executable files are not executable.

Possibly related this this bug introduced in alpine-3.14 -- [https://github.com/alpinelinux/docker-alpine/issues/156](https://github.com/alpinelinux/docker-alpine/issues/156)

## Expected behavior

bash-exec linter should be able to detect if files are executable or not.

## Steps to Reproduce

1) Launch super-linter in interactive mode
$ docker run -it --entrypoint bash github/super-linter

2) Create a test bash script, and validate that it is executable
bash-5.1# echo 'echo hello' > /tmp/test.sh && chmod +x /tmp/test.sh && stat /tmp/test.sh && /tmp/test.sh
  File: /tmp/test.sh
  Size: 11              Blocks: 8          IO Block: 4096   regular file
Device: 253,0   Inode: 582215      Links: 1
Access: (0755/-rwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-03-24 00:24:58.783723595 +0000
Modify: 2022-03-24 00:24:58.783723595 +0000
Change: 2022-03-24 00:24:58.784723595 +0000
 Birth: -
hello

3) run bash-exec linter against test bash script.
bash-5.1# /usr/bin/bash-exec /tmp/test.sh
Error: File:[/tmp/test.sh] is not executable

## Logs
```
bash-5.1# strace -s9999 -vvv -ff -T -ttt /usr/bin/bash-exec /tmp/test.sh
...
...
1648082256.627442 read(255, "#!/bin/bash \n\nif [[ -x \"$1\" ]]; then exit 0; else echo \"Error: File:[$1] is not executable\"; exit 1; fi", 103) = 103 <0.000008>
1648082256.627499 read(255, "", 103)    = 0 <0.000008>
1648082256.627547 mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ffff7eac000 <0.000008>
1648082256.627654 mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ffff7eab000 <0.000008>
1648082256.627721 munmap(0x7ffff7eab000, 4096) = 0 <0.000011>
1648082256.627767 mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ffff7eab000 <0.000008>
THIS LINE >>>>>>> 1648082256.627837 faccessat2(AT_FDCWD, "/tmp/test.sh", X_OK, AT_EACCESS) = -1 ENOSYS (Function not implemented) <0.000009>
1648082256.627922 munmap(0x7ffff7eab000, 4096) = -1 ENOSYS (Function not implemented) <0.000009>
1648082256.627965 mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = -1 ENOSYS (Function not implemented) <0.000035>
Error: File:[/tmp/test.sh] is not executable
1648082256.628053 writev(1, [{iov_base="Error: File:[/tmp/test.sh] is not executable", iov_len=44}, {iov_base="\n", iov_len=1}], 2) = -1 ENOSYS (Function not implemented) <0.000021>
1648082256.628143 rt_sigprocmask(SIG_BLOCK, [CHLD], 0x7fffffffe898, 8) = -1 ENOSYS (Function not implemented) <0.000009>
1648082256.628195 rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = -1 ENOSYS (Function not implemented) <0.000012>
1648082256.629099 +++ exited with 1 +++
```

**Comment by sabbene** [comment]
Maybe bash-exec can be replaced with something like:

`#!/bin/bash

if stat -c %A "${1}"  | grep -q x; then exit 0; else echo "Error: File:[$1] is not executable"; exit 1; fi`

**Comment by ferrarimarco** [comment]
Hi @sabbene . I didn't encounter this specific issue when super-linter checks scripts in a couple of repos (and the super-linter one as well).

Does it happen if you run super-linter against scripts that you didn't create inside the container?

**Comment by sabbene** [comment]
Hello @ferrarimarco,

Thanks for getting back to me.  I am able to repro this with a script created outside the container.

1) Create the script outside the container, and verify it is executable.
```$ echo 'echo hello' > /tmp/test.sh && chmod +x /tmp/test.sh && stat /tmp/test.sh && /tmp/test.sh
  File: ‘/tmp/test.sh’
  Size: 11              Blocks: 8          IO Block: 4096   regular file
Device: fd00h/64768d    Inode: 273521      Links: 1
Access: (0755/-rwxr-xr-x)  Uid: ( 9035/ sabbene)   Gid: (   30/    tape)
Access: 2022-03-24 09:13:00.612142468 -0700
Modify: 2022-03-24 09:17:45.008150367 -0700
Change: 2022-03-24 09:17:45.010150367 -0700
 Birth: -
hello```

2) Run super-linter in local mode
```
$ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -e LOG_LEVEL=WARN -v /tmp/test.sh:/tmp/lint/test.sh github/super-linter
...
...
2022-03-24 16:18:02 [WARN]   Warnings found in [bash-exec] linter!
2022-03-24 16:18:02 [WARN]   Error: File:[/tmp/lint/test.sh] is not executable
```

3) Verify that script does have executable bit set in container
```
$ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -e LOG_LEVEL=WARN -v /tmp/test.sh:/tmp/lint/test.sh -it --entrypoint bash github/super-linter
bash-5.1# ls -l /tmp/lint/test.sh
-rwxr-xr-x 1 9035 readproc 11 Mar 24 16:17 /tmp/lint/test.sh
```

**Comment by ferrarimarco** [comment]
Thanks for the additional tests. Can you try running `bash-exec` from inside the container against that script and report the return code? The invocation should be: `bash-exec path/to/script`

It would also be useful if you could run it (along with your tests) against one of the script that are bundled with super-linter, such as the ones in `/action/lib`. I'm trying to understand why our CI jobs didn't catch this.

**Comment by ferrarimarco** [comment]
You can see an example run here: https://github.com/github/super-linter/runs/5673252047?check_suite_focus=true#step:13:2686

**Comment by sabbene** [comment]
Here you go.  I ran against the same test file, and /action/lib/linter.sh

```
$ docker run -e RUN_LOCAL=true -e USE_FIND_ALGORITHM=true -e LOG_LEVEL=WARN -v /tmp/test.sh:/tmp/lint/test.sh -it --entrypoint bash github/super-linter
bash-5.1# /usr/bin/bash-exec /tmp/lint/test.sh
Error: File:[/tmp/lint/test.sh] is not executable
bash-5.1# echo $?
1

bash-5.1# /usr/bin/bash-exec /action/lib/linter.sh
Error: File:[/action/lib/linter.sh] is not executable
bash-5.1# echo $?
1
```

**Comment by ferrarimarco** [comment]
From https://github.com/alpinelinux/docker-alpine/issues/156#issuecomment-912645029, it appears it _might_ be related to the kernel version you're running. Can you post that as well?

The [GitHub-hosted runners run 5.11](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#ubuntu-20043-lts)

**Comment by ferrarimarco** [comment]
I could not reproduce it on a machine running a 5.15 kernel.

```shell
$ docker run --rm -it --entrypoint /bin/bash ghcr.io/github/super-linter                                                                                                                                     
bash-5.1# bash-exec /action/lib/linter.sh
bash-5.1# echo $?
0
```

**Comment by ferrarimarco** [comment]
> I see this in the example run you linked:
> 
> ```
> 2022-03-24 08:56:49 [DEBUG]   Setting LINTER_NAME to bash-exec...
> 2022-03-24 08:56:49 [DEBUG]   Setting LINTER_COMMAND to bash-exec...
> 2022-03-24 08:56:49 [DEBUG]   Setting FILE_ARRAY_VARIABLE_NAME to FILE_ARRAY_BASH_EXEC...
> 2022-03-24 08:56:49 [DEBUG]   FILE_ARRAY_BASH_EXEC file array contents: 
> 2022-03-24 08:56:49 [DEBUG]   Invoking bash-exec linter. TEST_CASE_RUN: false
> 2022-03-24 08:56:49 [DEBUG]    - No files found in changeset to lint for language:[BASH_EXEC]
> 2022-03-24 08:56:49 [DEBUG]   SKIP_FLAG: 1, list of files to lint: 
> ```
> 
> If I'm reading this right there aren't any bash files for it to test?

Ah I missed this. Search for `bash-exec`, you'll find some examples of that running correctly.

**Comment by ferrarimarco** [comment]
At least we know why :)

I'm going to close this because the 3.x Kernel it's not supported anymore. Thanks for helping us understand!

**Comment by sabbene** [comment]
Thats unfortunate, as the latest supported kernel from CentOS is 4.18 on CentOS8.  Nothing else that can be done?

**Comment by ferrarimarco** [comment]
How are you running super-linter? You could build a new image to override that file with your `stat`-based suggestion above, or you could bind-mount another implementation of the same (although this makes things less portable).

If it's running as an action in a self-hosted runner, maybe the lowest hanging fruit could be to disable that super-linter check by setting `VALIDATE_BASH_EXEC` to `false`, and add an ad-hoc step in the workflow that runs that custom script.

**Comment by sabbene** [comment]
Thanks for the suggestions.  I will give them a shot.
